hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientFlakeIdGeneratorProxy.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.proxy;
+
+import com.hazelcast.client.config.ClientFlakeIdGeneratorConfig;
+import com.hazelcast.client.impl.ClientMessageDecoder;
+import com.hazelcast.concurrent.flakeidgen.AutoBatcher;
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.codec.FlakeIdGeneratorNewIdBatchCodec;
+import com.hazelcast.client.impl.protocol.codec.FlakeIdGeneratorNewIdBatchCodec.ResponseParameters;
+import com.hazelcast.client.spi.ClientContext;
+import com.hazelcast.client.spi.ClientProxy;
+import com.hazelcast.client.spi.impl.ClientInvocation;
+import com.hazelcast.client.spi.impl.ClientInvocationFuture;
+import com.hazelcast.client.util.ClientDelegatingFuture;
+import com.hazelcast.core.FlakeIdGenerator;
+import com.hazelcast.core.IFunction;
+import com.hazelcast.core.IdBatch;
+import com.hazelcast.core.IdGenerator;
+import com.hazelcast.spi.InternalCompletableFuture;
+
+/**
+ * Proxy implementation of {@link IdGenerator}.
+ */
+public class ClientFlakeIdGeneratorProxy extends ClientProxy implements FlakeIdGenerator {
+
+    private static final ClientMessageDecoder NEW_ID_BATCH_DECODER = new ClientMessageDecoder() {
```

Comment:
You don't need a decoder if you're not returning the `InternalCompletableFuture`. Since you are `join`-ing on the future and returning an instance of the `IdBatch`, you can just as well do that in the `newIdBatch` or `newIdBatchAsync` method such as:
```
    @Override
    public IdBatch newIdBatch(int batchSize) {
        ClientMessage request = FlakeIdGeneratorNewIdBatchCodec.encodeRequest(name, batchSize);
        ClientInvocationFuture future = new ClientInvocation(getClient(), request, getName()).invoke();
        ResponseParameters response = FlakeIdGeneratorNewIdBatchCodec.decodeResponse(future.join());
        return new IdBatch(response.base, response.increment, response.batchSize);
    }
```
Decoders are good if you are going to return the future to the user. Am I correct @sancar ?